### PR TITLE
Add momentum and centered options to RMSProp

### PR DESCRIPTION
**Pull request by edouardelasalles**
Add two options :
 - Momentum (like SGD's momentum)
- Centered RMSprop, as in Graves 2013 ( https://arxiv.org/abs/1308.0850 ) : grad is normalized by running estimation of its variance

### CODE REVIEW

File: torch/optim/rmsprop.py [hunk 1]
```diff
@@ -49,7 +56,15 @@ def step(self, closure=None):
...
-                p.data.addcdiv_(-group['lr'], grad, avg)
+
+                avg = None
+                if group['centered']:
+                    grad_avg.mul_(alpha).add_(1 - alpha, grad)
+                    avg = square_avg.addcmul(-1, grad_avg, grad_avg).sqrt().add_(group['eps'])
+                else:
+                    avg = square_avg.sqrt().add_(group['eps'])
+
+                momentum.mul_(group['momentum']).addcdiv_(group['lr'], grad, avg)
```
Comment (apaszke) [inline torch/optim/rmsprop.py]: I think it's best to guard this line and the allocation of momentum buffer with an additional `if`. It will save some multiplications, but I think more importantly it will decrease the mem usage by the size of parameters.

**Review by apaszke — COMMENTED** [review]
There's the same problem with this PR and #774 - they will break already existing checkpoints because some options will be missing from parameter groups. I need to think what's the best way of fixing that.

**Comment by edouardelasalles** [comment]
Is this implementation better ?

**Comment by apaszke** [comment]
Yes, that's better. Can you please add a reference to that paper in the docstring? I'll get back to you tomorrow, with additional instructions on how to handle serialized optimizers. Thanks!

**Comment by apaszke** [comment]
Ok, so I thought this through and I think the best way would be to implement `__setstate__` such that it calls to super and then patches the param groups like that:
```python
def __setstate__(self, state):
    super(RMSprop, self).__setstate__(state)
    for group in self.parameter_groups:
        group.set_default('momentum', 0)
        group.set_default('centered', False)
```

File: torch/optim/rmsprop.py [hunk 2]
```diff
@@ -4,21 +4,35 @@
 class RMSprop(Optimizer):
     """Implements RMSprop algorithm.
 
+    Proposed by G. Hinton in its `course <http://www.cs.toronto.edu/~tijmen/csc321/slides/lecture_slides_lec6.pdf>`_.
```
Comment (apaszke) [inline torch/optim/rmsprop.py]: in `his` course

File: torch/optim/rmsprop.py [hunk 3]
```diff
@@ -49,7 +67,20 @@ def step(self, closure=None):
...
+                if group['centered']:
+                    grad_avg = state['grad_avg']
+                    grad_avg.mul_(alpha).add_(1 - alpha, grad)
+                    avg = square_avg.addcmul(-1, grad_avg, grad_avg).sqrt().add_(group['eps'])
+                else:
+                    avg = square_avg.sqrt().add_(group['eps'])
+
+                if group['momentum'] > 0:
+                    mom = state['momentum']
+                    mom.mul_(group['momentum']).addcdiv_(group['lr'], grad, avg)
```
Comment (apaszke) [inline torch/optim/rmsprop.py]: So my only concern here is that e.g. in `sgd` the momentum is multiplied by the learning rate only when it is added the the parameters. It seems to me that it doesn't change the formula (right?), and I think it's better to keep like that to be consistent.

**Review by apaszke — COMMENTED** [review]
Where does the "centered" come from? It doesn't seem to be called like that in the paper you linked.

**Comment by edouardelasalles** [comment]
I do not know where the  term "centered" came from. I read it in several DeepMind's papers (A3C and Recurrent Environnement Simulator). And it is the denomination given in TensorFlow's implementation.

File: torch/optim/rmsprop.py (hunk 3 above)
Comment (edouardelasalles) [inline torch/optim/rmsprop.py]: In the orignal momentum algorithm, the learning rate is multiply by the gradient before the momentum is applied.

Comment (ajbrock) [inline torch/optim/rmsprop.py]: Applying the LR to the momentum as is done here is equivalent to moving the multiply outside those terms (and holds for Nesterov momentum as well), though I'm not sure if floating-point precision becomes an issue for small learning rates.

Comment (apaszke) [inline torch/optim/rmsprop.py]: Yeah it's equivalent, but we decided that it's best to keep it consistent. The alternative formulation has the advantage that if you modify the LR, it is immediately reflected in the updates, instead of slowly making the buffer smaller.

File: torch/optim/rmsprop.py [hunk 4]
```diff
@@ -49,7 +67,20 @@ def step(self, closure=None):
                     grad = grad.add(group['weight_decay'], p.data)
 
                 square_avg.mul_(alpha).addcmul_(1 - alpha, grad, grad)
-                avg = square_avg.sqrt().add_(group['eps'])
-                p.data.addcdiv_(-group['lr'], grad, avg)
+
+                avg = None
```
Comment (apaszke) [inline torch/optim/rmsprop.py]: No need to set it to `None`. Variables declared inside the `if` blocks are visible outside too.

File: torch/optim/rmsprop.py [hunk 5]
```diff
@@ -49,7 +67,20 @@ def step(self, closure=None):
...
+                if group['centered']:
+                    grad_avg = state['grad_avg']
+                    grad_avg.mul_(alpha).add_(1 - alpha, grad)
+                    avg = square_avg.addcmul(-1, grad_avg, grad_avg).sqrt().add_(group['eps'])
+                else:
+                    avg = square_avg.sqrt().add_(group['eps'])
+
+                if group['momentum'] > 0:
+                    mom = state['momentum']
+                    mom.mul_(group['momentum']).addcdiv_(grad, avg)
```
Comment (apaszke) [inline torch/optim/rmsprop.py]: can you use a different name for momentum parameter and buffer? It's quite confusing. Just rename it to `momentum_buffer` (that's what we use in SGD).

File: torch/optim/rmsprop.py (hunk 4 above)
Comment (edouardelasalles) [inline torch/optim/rmsprop.py]: Yes, it is only for readability, but i'll remove it.

**Review by apaszke — APPROVED** [review]

File: torch/optim/rmsprop.py [hunk 6]
```diff
@@ -4,21 +4,35 @@
...
             numerical stability (default: 1e-8)
+        centered (bool, optional) : if True, compute the centered RMSProp,
+            the gradient is normalized by an estimation of its variance
         weight_decay (float, optional): weight decay (L2 penalty) (default: 0)
 
     """
 
-    def __init__(self, params, lr=1e-2, alpha=0.99, eps=1e-8, weight_decay=0):
-        defaults = dict(lr=lr, alpha=alpha, eps=eps, weight_decay=weight_decay)
+    def __init__(self, params, lr=1e-2, momentum=0, alpha=0.99, eps=1e-8, centered=False, weight_decay=0):
```
Comment (apaszke) [inline torch/optim/rmsprop.py]: Ugh, sorry but I just realised that you inserted additional arguments at the beginning of `__init__`. We can't do that, because if someone created the optimizer like that:
```
optimizer.RMSprop(model.parameters(), 1e-2, 1 - 1e-4)
```
then the third argument will be interpreted as momentum, while it used to be `alpha`. Once that is change it's good to be merged.

**Comment by apaszke** [comment]
Thank you!
